FUSETOOLS-1856: Allow to specify what kind of information will be displayed in camel components

apodhrad opened the pull request.

apodhrad commented:
All sonar complaints were fixed but the PR can be merged in the next sprint

lhein commented:
It is a new feature and therefore we are not going to merge it in 10.1. It will be part of 10.2.

apupier commented on core/tests/org.fusesource.ide.camel.model.service.core.tests.integration/src/main/java/org/fusesource/ide/camel/model/service/core/tests/integration/core/model/CamelBasicModelElementTestIT.java:
```diff
@@ -0,0 +1,61 @@
+	@Before
+	@After
+	public void cleanUserLabelsPreference() {
+		PreferenceManager.getInstance().savePreference(EDITOR_PREFERRED_LABEL, "");
```
you are nto sure to let the systeùm in teh same state like that.
You need to retrieve the value in the @Before and then set it back in @After

apupier commented on core/tests/org.fusesource.ide.camel.model.service.core.tests.integration/src/main/java/org/fusesource/ide/camel/model/service/core/tests/integration/core/model/CamelBasicModelElementTestIT.java:
```diff
@@ -0,0 +1,61 @@
+		assertEquals("Log ${body}", simpleLog().getDisplayText());
+	}
+
+	private static AbstractCamelModelElement simpleLog() {
```
There is no need to use a static method. I would prefer to avoid static usage when not necessary

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/preferences/ComponentValidator.java:
```diff
@@ -0,0 +1,54 @@
+		if (newText.isEmpty()) {
+			return UIMessages.preferredLabelsErrorMessageEmptyComponent;
+		}
+		if (newText.matches(".*[\\W&&[^-]].*")) {
```
why it is not the same than org.fusesource.ide.camel.model.service.core.model.AbstractCamelModelElement.USER_LABEL_REGEX  "[\\w-]+\\.[\\w-]+" ?

apodhrad replied:
these two regular expressions are different but I will refactor it so that we can use constants from that class.

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/preferences/ParameterValidator.java:
```diff
@@ -0,0 +1,34 @@
+		if (newText.isEmpty()) {
+			return UIMessages.preferredLabelsErrorMessageEmptyParameter;
+		}
+		if (newText.matches(".*[\\W&&[^-]].*")) {
```
it is the same than org.fusesource.ide.camel.editor.preferences.ComponentValidator.isValid(String), what about factorizing it?

apodhrad replied:
these two regular expressions are different but I will refactor it so that we can use constants from that class.

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/preferences/PreferredLabelDialog.java:
```diff
@@ -0,0 +1,219 @@
+	protected Control createDialogArea(Composite parent) {
+		Composite area = (Composite) super.createDialogArea(parent);
+		Composite container = new Composite(area, SWT.NONE);
+		container.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, true));
```
just FYI, i usually prefer to use GridDataFactory.fillDefaults().grab(true, true).create() it avoids me to remind about all the parameters but using the n"normal" constructor is perfectly fine.

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/preferences/PreferredLabelDialog.java:
```diff
@@ -0,0 +1,219 @@
+	}
+
+	protected List<Eip> getComponents() {
+		return getCamelModel().getEips().stream().filter(eip -> !"NONE".equals(getCategoryFromEip(eip)))
```
would be nice to extract a constant for NONE and reuse it in ProviderHelper

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/preferences/PreferredLabelEditor.java:
```diff
@@ -0,0 +1,410 @@
+	private Button removeButton;
+
+	/**
+	 * The selection listener.
```
This javadoc is not providing more help than the name of the attributes itself. Either provide more details or please remove it

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/preferences/PreferredLabelEditor.java:
```diff
@@ -0,0 +1,410 @@
+	}
+
+	/**
+	 * Returns this field editor's selection listener. The listener is created if nessessary.
```
typo nessessary -> necessary

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/preferences/PreferredLabelEditor.java:
```diff
@@ -0,0 +1,410 @@
+
+	@Override
+	protected void doStore() {
+		String s = createList(getPreferredLabels());
```
what is "s"? The String representation of all the "preferredLabels"?
Can you provide a more meaningful name please.

apupier commented on editor/tests/org.fusesource.ide.camel.editor.tests.integration/META-INF/MANIFEST.MF:
```diff
@@ -30,6 +30,7 @@ Require-Bundle: org.junit;bundle-version="4.12.0",
  org.fusesource.ide.camel.model.service.core.tests.integration,
  org.fusesource.ide.branding;bundle-version="10.0.0",
  org.fusesource.ide.projecttemplates;bundle-version="10.0.0",
+ org.fusesource.ide.preferences;bundle-version="10.0.0",
```
I think that it is 10.2 (even if we are not taking a great care about it)

apodhrad replied:
all the versions should be updated in one commit

lhein replied:
I saw several manifests pointing to 10.0.0

apupier replied:
@apodhrad what do you mean exactly  by "it should be updated in one commit"? We should update all of them when upgrading Fuse Tooling version?

As I said previously, we are not taking great care of these requirements version as we always install against the same Target Platform and with All Fuse Tooling plugins of the same version.

I think it is better to use the latest version when adding a new one as we are not checking backward-compatibility.

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/internal/l10n/messages.properties:
```diff
@@ -215,3 +215,20 @@ selectEndpointWizardWindowTitle=Select component
+preferredLabelsComponentHeader=Component 
+preferredLabelsParameter=Parameter:  
+preferredLabelsParameterHeader=Parameter
+preferredLabelsErrorMessage=Input is not valid
```
Maybe worth precising why it is not valid

apodhrad replied:
This message is not used, so I will remove it.

apodhrad commented:
@apupier I have applied some of your suggestions, the others can be applied in other PRs.

apodhrad commented:
testPR

bfitzpat commented:
Thanks for the demo today @apodhrad ! Looks great. 

A few thoughts:

1. Would be great to have a way to do more global changes. So if there's a "description" attribute and you want that to show for everything possible in the diagram, you could set that. And then override that by layering additional changes -- like perhaps you want the log to show the URI instead. 
2. Would be nice to be able to have global preferences as well as project-specific overrides. 
3. We should simplify how the user can share these preferences across workspaces, or at least document how they can be shared. 

But as a first pass, this looks like a great addition and should provide a good level of customization for our users. Great work!

jpav commented:
Another suggestion would be to maybe have a right-click (i.e., context) menu that allowed the user to set the display info.

apupier commented on core/plugins/org.fusesource.ide.camel.model.service.core/src/org/fusesource/ide/camel/model/service/core/model/AbstractCamelModelElement.java:
```diff
@@ -57,13 +57,20 @@
 	public static final String WIRETAP_NODE_NAME = "wireTap";
+
+	public static final String USER_LABEL_COMPONENT_REGEX = "[\\w-]+";
+	public static final String USER_LABEL_PARAMETER_REGEX = "[\\w-]+";
```
Do you want to keep this 2 variables separated? or maybe it can be merged in a USER_LABEL_REGEXP ?

apodhrad replied:
yes, at the moment the expressions are the same but this may be changed in the future with high probability

apodhrad replied:
but I can follow sonar suggestion and add a comment

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/preferences/PreferredLabelDialog.java:
```diff
@@ -0,0 +1,218 @@
+	}
+
+	protected List<Eip> getComponents() {
+		return getCamelModel().getEips().stream().filter(eip -> belongsToAnyCategory(eip)).collect(toList());
```
Can you replace follow the Sonar recommendation please?

apodhrad replied:
ok

apodhrad commented:
alerady rebased

apodhrad commented:
Hey guys, thank you for your suggestions. I have already created Jira issues for them so that they can be fixed in other pull requests.

apodhrad commented:
There is one failing test but this test has nothing in common with this feature.

apodhrad commented:
@lhein @apupier could you have a look at it, please? The feature freeze is coming ;-)

lhein commented:
testPR

lhein commented:
What is with the critical sonar issue? is that fixed already? Can't see it.

apodhrad commented:
@lhein yes, I have fixed them

lhein commented on core/tests/org.fusesource.ide.camel.model.service.core.tests.integration/META-INF/MANIFEST.MF:
```diff
@@ -16,9 +16,10 @@ Require-Bundle: org.eclipse.core.filesystem,
  org.fusesource.ide.foundation.ui;bundle-version="10.0.0",
- org.fusesource.ide.camel.model.service.core,
- org.fusesource.ide.camel.model.service.impl,
+ org.fusesource.ide.camel.model.service.core;bundle-version="10.0.0",
```
is there any reason for requiring a specific version ?

apodhrad replied:
The tests will not work with bundles 9.x

lhein commented:
I still see some points open from Aurelien which should be fixed first.

apodhrad commented:
@apupier @lhein is there any other requirement which must be implemented in this PR?

apupier commented:
@apodhrad from the list https://github.com/jbosstools/jbosstools-fuse/pull/1133#pullrequestreview-76416431
I think that before merging we should implement:
- when choosing a parameter which has a default value, it is displayed only if the parameter is not the default one (or has been already changed in UI)
--> https://issues.jboss.org/browse/FUSETOOLS-2707
- when I remove a "preferred label" and that i click Apply and close, there is no id label displayed until I switch tabs of the design editor
--> https://issues.jboss.org/browse/FUSETOOLS-2706

as they are more bugs than "missing feature"/"enhancement"

and
- in the "New Preferred Label" dialog, it would be nice to add a description to the dialog to avoid the blank. Something like: "Select the parameter value to display in diagram for a specific Camel Component"
--> https://issues.jboss.org/browse/FUSETOOLS-2705

as it is a very simple thing which greatly improves the usability

apodhrad commented:
@apupier thx for the response, I'll do it asap

apodhrad commented:
Fixed the issue with default values and added description to the dialog. But I'm not able to reproduce your issue with refreshing after clicking on 'Apply and Close', see https://issues.jboss.org/browse/FUSETOOLS-2706

apupier commented:
ensure that there is a green build but I think it is fine for a first iteration.

I noticed another little thing that can be done in another PR with the usability improvements that we are doing.
![image](https://user-images.githubusercontent.com/1105127/36597310-07e6acd2-18a9-11e8-8148-1de07f3a58fa.png)
the validation message is only at the top of the Wizard. It would be nice to have it also on the field that needs modifications too.

apodhrad commented:
testPR

lhein commented:
+1 once we have a green PR build

apodhrad commented:
testPR

apodhrad commented:
testPR

lhein commented:
testPR
